backend/handler/passcode.go

```diff
@@ -231,11 +261,31 @@ func (h *PasscodeHandler) Finish(c echo.Context) error {
 			return fmt.Errorf("failed to delete passcode: %w", err)
 		}
 
-		if !user.Verified {
-			user.Verified = true
-			err = userPersister.Update(*user)
+		if passcode.Email.User != nil && passcode.Email.User.ID.String() != user.ID.String() {
+			return dto.NewHTTPError(http.StatusForbidden, "email address has been claimed by another user")
+		}
+
+		if !passcode.Email.Verified {
+			// Update email verified status and assign the email address to the user.
+			passcode.Email.Verified = true
+			passcode.Email.UserID = &user.ID
+
+			err = emailPersister.Update(passcode.Email)
+			if err != nil {
+				return fmt.Errorf("failed to update the email verified status: %w", err)
+			}
+
+			err = h.auditLogger.Create(c, models.AuditLogEmailVerified, user, nil)
 			if err != nil {
-				return fmt.Errorf("failed to update user: %w", err)
+				return fmt.Errorf("failed to create audit log: %w", err)
+			}
+
+			if user.Emails.GetPrimary() == nil {
+				primaryEmail := models.NewPrimaryEmail(passcode.Email.ID, user.ID)
+				err = primaryEmailPersister.Create(*primaryEmail)
+				if err != nil {
+					return fmt.Errorf("failed to create primary email: %w", err)
+				}
```

Comment:
I think we should also add an audit log here, that the primary email was added (or changed).